app/code/Magento/Catalog/Model/ResourceModel/Product/Indexer/Eav/Source.php

```diff
@@ -297,6 +308,47 @@ protected function _prepareMultiselectIndex($entityIds = null, $attributeId = nu
         return $this;
     }
 
+    /**
+     * Get options for multiselect attributes using custom source models
+     * @maderlock's fix from: https://github.com/magento/magento2/issues/417#issuecomment-265146285
+     *
+     * @param array $attrIds
+     * @param array $options
+     *
+     * @return array
+     * @throws \Zend_Db_Statement_Exception
+     */
+    protected function _getMultiSelectAttributeWithSourceModels($attrIds, $options)
+    {
+        // Add options from custom source models
+        $select = $this->getConnection()->select()
+            ->from(
+                ['ea' => $this->getTable('eav_attribute')],
+                ['attribute_id','entity_type_id', 'attribute_code']
+            )
+            ->where('attribute_id IN(?)', $attrIds)
+            ->where('source_model is not null');
+        $query = $select->query();
+
+        while ($row = $query->fetch()) {
+            try {
+                /** @var \Magento\Catalog\Model\ResourceModel\Eav\Attribute $attribute */
+                $attribute = $this->_attributeRepository->get($row['entity_type_id'], $row['attribute_code']);
+                $sourceModelOptions = $attribute->getOptions();
+                // Add options to list used below
+                foreach ($sourceModelOptions as $o) {
+                    $options[$row['attribute_id']][$o->getValue()] = true;
+                }
+            } catch (\BadMethodCallException $e) {
+                // Skip
+            } catch (\Magento\Framework\Exception\NoSuchEntityException $e) {
```

Comment:
@sivaschenko The `NoSuchEntityException` is to handle the potential of `$this->_attributeRepository->get()` throwing such an error. Would you recommend just logging an exception? 
The `BadMethodCallException` I think should be removed actually. I believe it was left over from the original code using `ObjectManager`. I don't see where that would get thrown anymore, so that's my bad. I can remove it and add some logging for the other.

Reply:
@rain2o thanks for the clarification, yes I do recommend logging NoSuchEntityException as the source of attribute information is internal and in this situation a missing attribute is an exceptional case

Reply:
@sivaschenko Due to the refactor to use `getList` this function no longer throws any exceptions that I'm aware of, so I have removed all try/catch blocks.